addons/binding/org.openhab.binding.gpstracker/src/main/java/org/openhab/binding/gpstracker/internal/handler/ChannelUtil.java

```diff
@@ -0,0 +1,267 @@
+
+    /**
+     * Thing representing the tracker tracker.
+     */
+    private Thing thing;
+
+    /**
+     * Binding level configuration
+     */
+    private BindingConfiguration bindingConfig;
+
+    /**
+     * Tracker level configuration
+     */
+    private TrackerConfiguration trackerConfig;
+
+    /**
+     * Translation helper
+     */
+    private TranslationUtil translationUtil;
+
+
+    /**
+     * Constructor.
+     *
+     * @param thing           The thing
+     * @param bindingConfig   Binding level configuration
+     * @param trackerConfig   Device level configuration
+     * @param translationUtil Translation helper
+     */
+    ChannelUtil(Thing thing, BindingConfiguration bindingConfig, TrackerConfiguration trackerConfig, TranslationUtil translationUtil) {
+        this.thing = thing;
+        this.bindingConfig = bindingConfig;
+        this.trackerConfig = trackerConfig;
+        this.translationUtil = translationUtil;
+    }
+
+    void setTrackerConfig(TrackerConfiguration trackerConfig) {
+        this.trackerConfig = trackerConfig;
+    }
+
+    /**
+     * Updates dynamic channels for the .
+     * - A Switch type channel will be created for all binding and tracker level regions if the region triggerEvent
+     * flag is FALSE. Channel state will be set to ON if the tracker is inside the region.
+     * - RegionEnter and regionLeave trigger channels will be created if triggerEvent flag for region is TRUE. The
+     * payload for these channels will be the region name.
+     */
+    Map<String, Channel> updateChannels() {
+        ThingUID uid = thing.getUID();
+        logger.debug("Update channel structure based on binding configuration for tracker {}", uid.getId());
+        Map<String, Channel> channelMap = getChannelMap();
+        Set<String> origChannels = new HashSet<>(channelMap.keySet());
+
+        removeUnusedChannels(channelMap);
+
+        createMissingChannels(channelMap);
+
+        if (logger.isTraceEnabled()) {
+            for (Channel c : channelMap.values()) {
+                logger.trace("Channel after update: {}", c.getUID().getId());
+            }
+        }
+
+        Set<String> newChannels = new HashSet<>(channelMap.keySet());
+        return newChannels.equals(origChannels) ? null : channelMap;
+    }
+
+    /**
+     * Create missing region switches and trigger channels.
+     *
+     * @param channelMap Channel map with required thing channels.
+     */
+    private void createMissingChannels(Map<String, Channel> channelMap) {
+        ThingUID uid = thing.getUID();
+        boolean triggerChannelsNeeded = false;
+        for (Region r : getAllRegions()) {
+            String presenceChannelId = CHANNEL_REGION_PRESENCE + "_" + r.getId();
+            if (!r.getTriggerEvent()) {
+                if (!channelMap.containsKey(presenceChannelId)) {
+                    Channel channel = ChannelBuilder.create(new ChannelUID(uid, presenceChannelId),
```

Comment:
@cweitkamp The reason I would like to customize channel names is that I think its more user friendly:
![Image](http://gdurl.com/7UfW)
Of course you are the maintainers - if it's against some policy I'll sadly push the poor version... :disappointed

Reply:
@kaikreuzer I was just typing my answer :wink:.

@gbicskei I guess we have to clarify terms first. What does "channel name" mean to you? We have a channel id, a label and a description.

Please correct me if I am wrong. Currently you append the user defined region to the channel id (e.g. `"presence" + "_" + regionId`). From my point of view that is critical because the channel id is limited to the following pattern: `[\\w-]*|[\\w-]*#[\\w-]`. How do you handle characters which are not allowed?

I am sorry if my questions seeded confusion. I had in my mind, that using the `BundleContext` in bindings is not allowed. But that was wrong. As @kaikreuzer stated in https://github.com/openhab/openhab2-addons/pull/3839#discussion_r214756982 it is okay to pass it from the HandlerFactory to the `TranslationUtil` class. This means your approach on how to translate the channels labels is fine and my proposed discussion is not necessary.

If you have an idea how we can make it easier in the framework please come up with a contribution for it or feel free to contact me.

Reply:
@cweitkamp 
> that is critical because the channel id is limited to the following pattern: [\\w-]*|[\\w-]*#[\\w-]. How do you handle characters which are not allowed?

The current solution filters the Id for Regions:
```
public String getId() {
        return name != null ? name.replaceAll("[^a-zA-Z0-9_]", ""): "UnnamedRegion";
    }
```

Reply:
@cweitkamp 
You are right, I was not precise enough. By channel name I meant the label that is showed on the screenshot above.